src/app/util/attribute-storage.h

```diff
@@ -235,6 +235,40 @@ const EmberAfCluster * emberAfFindClusterInType(const EmberAfEndpointType * endp
 // Initial configuration
 void emberAfEndpointConfigure();
 
+// setup a dynamic endpoint's EmberAfEndpointType from a list of template clusters.
+//
+// This is a alternative to declaring dynamic endpoint metadata using DECLARE_DYNAMIC_* macros.
+//
+// As clusters to be used in dynamic endpoint setup need to be defined in ZAP anyway
+// (usually on a special endpoint which remains always disabled), the cluster's
+// metadata including all attributes already exists and can be re-used this way,
+// without error prone manual duplicating with DECLARE_DYNAMIC_*
+//
+// templateEndpointId specifies a endpoint which is usually disabled, but containing
+// cluster definitions that should be used for instantiating active endpoints.
+//
+// templateClusterIds is a list of clusterIds to be used for the new endpoint.
+//
+// endpointType will be setup with the specified clusters and their storage size so
+// it can be used in a subsequent call to emberAfSetDynamicEndpoint instead of
+// an endpoint manually constructed with DECLARE_DYNAMIC_*.
+//
+// Note: passing invalid templateEndpointId/templateClusterIds combinations, i.e. clusters
+//   not present in the specified template endpoint, will cause the function will die with
+//   an error message as this indicates severe internal inconsistency of the setup.
+//
+// Note: function may allocate memory for the endpoint declaration.
+//   Use emberAfResetEndpointDeclaration to properly dispose of an endpoint declaration.
+void emberAfSetupDynamicEndpointDeclaration(EmberAfEndpointType & endpointType, chip::EndpointId templateEndpointId,
```

Comment:
should we make endpointType be verifiable empty so that we avoid memory leaks if I do `emberAfSetupDynamicEndpointDeclaration` twice on the same endpoint without a reset?

Reply:
How could that be done? We'd need a flag differentiating dynamically set up endpoint types from those generated via ZAP/codegen. And as long as `EmberAfEndpointType` is a simple struct, any members, including a potential flag, are undefined before the endpoint is passed to a initializer function such as `emberAfSetupDynamicEndpointDeclaration`.

Reply:
You have code of `endpointType.cluster = new EmberAfCluster[endpointType.clusterCount];` which replaces any `.cluster` there.

I was thinking to enforce that `endpointType.cluster == nullptr` as a prerequisite for this function.